File: mobile_app/lib/services/employee_service.dart
import '../config/api_config.dart';
import '../models/employee_model.dart';
import 'api_service.dart';

class EmployeeService {
  final ApiService _apiService = ApiService();

  // Get all employees
  Future<List<Employee>> getAllEmployees() async {
    try {
      final response = await _apiService.get(ApiConfig.employeesEndpoint);
      final data = _apiService.handleResponse(response) as List;
      
      return data.map((json) => Employee.fromJson(json)).toList();
    } catch (e) {
      throw Exception('Failed to load employees: $e');
    }
  }

  // Get employees by department
  Future<List<Employee>> getEmployeesByDepartment(String departmentId) async {
    try {
      final response = await _apiService.get(
        '${ApiConfig.departmentsEndpoint}/$departmentId/employees',
      );
      final data = _apiService.handleResponse(response) as List;
      
      return data.map((json) => Employee.fromJson(json)).toList();
    } catch (e) {
      throw Exception('Failed to load employees: $e');
    }
  }

  // Get employee by ID
  Future<Employee> getEmployeeById(String departmentId, String employeeId) async {
    try {
      final response = await _apiService.get(
        '${ApiConfig.departmentsEndpoint}/$departmentId/employees/$employeeId',
      );
      final data = _apiService.handleResponse(response);
      
      return Employee.fromJson(data);
    } catch (e) {
      throw Exception('Failed to load employee: $e');
    }
  }

  // Create employee
  Future<Employee> createEmployee(String departmentId, Employee employee) async {
    try {
      final response = await _apiService.post(
        '${ApiConfig.departmentsEndpoint}/$departmentId/employees',
        employee.toJson(),
      );
      final data = _apiService.handleResponse(response);
      
      return Employee.fromJson(data);
    } catch (e) {
      throw Exception('Failed to create employee: $e');
    }
  }

  // Update employee
  Future<Employee> updateEmployee(
    String departmentId,
    String employeeId,
    Employee employee,
  ) async {
    try {
      final response = await _apiService.put(
        '${ApiConfig.departmentsEndpoint}/$departmentId/employees/$employeeId',
        employee.toJson(),
      );
      final data = _apiService.handleResponse(response);
      
      return Employee.fromJson(data);
    } catch (e) {
      throw Exception('Failed to update employee: $e');
    }
  }

  // Delete employee
  Future<void> deleteEmployee(String departmentId, String employeeId) async {
    try {
      await _apiService.delete(
        '${ApiConfig.departmentsEndpoint}/$departmentId/employees/$employeeId',
      );
    } catch (e) {
      throw Exception('Failed to delete employee: $e');
    }
  }
}

File: backend/src/main/java/com/enterprise/department/service/EmployeeService.java
package com.enterprise.department.service;

import com.enterprise.department.dto.EmployeeDTO;

import java.util.List;
import java.util.Map;

/**
 * Employee Service Interface
 */
public interface EmployeeService {

    /**
     * Get all employees
     *
     * @return List of employees
     */
    List<EmployeeDTO> getAllEmployees();

    /**
     * Get employees by department ID
     *
     * @param departmentId Department ID
     * @return List of employees
     */
    List<EmployeeDTO> getEmployeesByDepartment(String departmentId);

    /**
     * Get employee by ID
     *
     * @param id Employee ID
     * @return EmployeeDTO
     */
    EmployeeDTO getEmployeeById(String id);

    /**
     * Create new employee
     *
     * @param departmentId Department ID
     * @param employeeDTO  Employee data
     * @return Created employee
     */
    EmployeeDTO createEmployee(String departmentId, EmployeeDTO employeeDTO);

    /**
     * Update existing employee
     *
     * @param departmentId Department ID
     * @param employeeId   Employee ID
     * @param employeeDTO  Updated employee data
     * @return Updated employee
     */
    EmployeeDTO updateEmployee(String departmentId, String employeeId, EmployeeDTO employeeDTO);

    /**
     * Delete employee
     *
     * @param departmentId Department ID
     * @param employeeId   Employee ID
     */
    void deleteEmployee(String departmentId, String employeeId);

    /**
     * Group employees by department
     * Demonstrates use of Maps and Lists
     *
     * @return Map of department ID to list of employees
     */
    Map<String, List<EmployeeDTO>> groupEmployeesByDepartment();

    /**
     * Calculate total salary by department
     * Demonstrates use of Streams API
     *
     * @return Map of department ID to total salary
     */
    Map<String, Double> calculateTotalSalaryByDepartment();
}
